### ISSUE TITLE
Helm pull does not use cache

### ISSUE THREAD

**Issue by olivierboudet** [helm/helm#12602]
Helm pull does not use local cache as other commands do. It fetch the chart and saves it in current directory (or directory specified with --destination) but does not keep it in cache. So it is downloaded each time we try to fetch it.

I think we should keep it in cache and just copy it to the wanted destination.

How to reproduce : 

* check that chart does not exist in cache :
```
ls -l ~/.cache/helm/repository/mongodb-14.3.0.tgz
"/home/olivier/.cache/helm/repository/mongodb-14.3.0.tgz": No such file or directory (os error 2)
```
* run helm pull : `helm pull bitnami/mongodb --version 14.3.0`
* check that chart does not exist in cache : 
```
ls -l ~/.cache/helm/repository/mongodb-14.3.0.tgz
"/home/olivier/.cache/helm/repository/mongodb-14.3.0.tgz": No such file or directory (os error 2)
```
* run helm template : `helm template bitnami/mongodb`
* check that chart exists in cache :
```
ls -l ~/.cache/helm/repository/mongodb-14.3.0.tgz
.rw-r--r-- 83k olivier 29 nov.  17:40 /home/olivier/.cache/helm/repository/mongodb-14.3.0.tgz
```

Output of `helm version`:

version.BuildInfo{Version:"v3.13.2", GitCommit:"2a2fb3b98829f1e0be6fb18af2f6599e0f4e8243", GitTreeState:"clean", GoVersion:"go1.20.10"}

**Comment by Ithrael** [comment]
I can reproduce this issue. 
I believe the 'helm template' command is more focused on local operation, thus there is a need for caching. On the other hand, the original intent of 'helm pull' should be to fetch charts from the remote repository; considering this aspect, I think it is reasonable for 'helm pull' not to cache. Perhaps we could add an option to 'helm pull'? For example: 'helm pull xxx --cache'.

**Comment by olivierboudet** [comment]
Yes, one of option `--cache` or `--no-cache` depending on which behavior should be the default seems reasonable to me.

**Comment by Ithrael** [comment]
I originally wanted to add cache functionality for helm pull. If I specify --cache, helm pull will cache the chart in the repositoryCache directory and download the chart to a specified directory. Moreover, if there is already a chart in the repositoryCache directory, it will not download it, just copy it to the specified directory. Later I noticed that although helm template can cache, if the repositoryCache already has a chart, running helm template again will still re-download the chart and cache it in the repositoryCache directory. 

related: https://github.com/helm/helm/issues/12607

**Comment by olivierboudet** [comment]
> I originally wanted to add cache functionality for helm pull. If I specify --cache, helm pull will cache the chart in the repositoryCache directory and download the chart to a specified directory. Moreover, if there is already a chart in the repositoryCache directory, it will not download it, just copy it to the specified directory. Later I noticed that although helm template can cache, if the repositoryCache already has a chart, running helm template again will still re-download the chart and cache it in the repositoryCache directory. 
> 
> related: https://github.com/helm/helm/issues/12607
> 

Yes I observed same behavior for helm template when trying to do a PR for this issue.

**Comment by moritzschmitz-oviva** [comment]
Is it possible to implement this via plugin perhaps?

**Comment by TerryHowe** [comment]
Are you talking about a cache with helm repositories or oci registries?

**Comment by cyberhck** [comment]
please remove stale label from here, it's still relevant. and now that the oras-project/oras-go#1086 is merged, what's next? how do we get this done?

**Comment by TerryHowe** [comment]
It is going to be a while before all the changes to oras-go are complete and Helm utilizes them.

**Comment by cyberhck** [comment]
> Are you talking about a cache with helm repositories or oci registries?

My usecase is oci registries, I understand I can do a wrapper and then have it download and use that path as well, but I feel it belongs in helm

**Comment by cyberhck** [comment]
If someone could point me to a direction where I could try to do caching while using OCI registries, I'd love to give this a stab.

**Comment by TerryHowe** [comment]
I think it would be a better use of your time to help getting the oras-go upgrade done. 
* https://github.com/helm/helm/pull/32065
* https://github.com/oras-project/oras-go

**Comment by cyberhck** [comment]
absolutely! how can I help? I can see there are some good first issue in the list. Is there something that's easy to get started and also helps you with using the oras-go?

**Comment by TerryHowe** [comment]
@cyberhck one possible way to help would be take small features from https://github.com/oras-project/oras-go/pull/1130 and create PRs for them to main. If that sounds possible I could probably give you some ideas.

**Comment by cyberhck** [comment]
hey, sorry for being back quite late, I was on holiday at the moment), just seeing this,

I've read that issue, from the overview I don't know what it entails, being new to the repo, the simplest seems to be  to try logging & transport section's `LoggingTransport for HTTP request/response debug logging` or `Logger field on ClientBuilder` I could help on that if you can brief me on that.
